src/hotspot/share/opto/library_call.cpp

```diff
@@ -3715,6 +3718,102 @@ bool LibraryCallKit::inline_native_setScopedValueCache() {
+  Node* last_continuation = make_load(control(), last_continuation_offset, last_continuation_offset->get_ptr_type(), T_ADDRESS, MemNode::unordered);
+
+  // Null check the last continuation object.
+  Node* continuation_cmp_null = _gvn.transform(new CmpPNode(last_continuation, null()));
+  Node* test_continuation_not_equal_null = _gvn.transform(new BoolNode(continuation_cmp_null, BoolTest::ne));
+  IfNode* iff_continuation_not_equal_null = create_and_map_if(control(), test_continuation_not_equal_null, PROB_MAX, COUNT_UNKNOWN);
+
+  // False path, last continuation is null.
+  Node* continuation_is_null = _gvn.transform(new IfFalseNode(iff_continuation_not_equal_null));
+
+  // True path, last continuation is not null.
+  Node* continuation_is_not_null = _gvn.transform(new IfTrueNode(iff_continuation_not_equal_null));
+
+  set_control(continuation_is_not_null);
+
+  // Load the pin count from the last continuation.
+  Node* pin_count_offset = basic_plus_adr(top(), last_continuation, in_bytes(ContinuationEntry::pin_count_offset()));
+  Node* pin_count = make_load(control(), pin_count_offset, TypeInt::INT, T_INT, MemNode::unordered);
+
+  // The loaded pin count is compared against a context specific rhs for over/underflow detection.
+  Node* pin_count_rhs;
+  if (unpin) {
+    pin_count_rhs = _gvn.intcon(0);
+  } else {
+    pin_count_rhs = _gvn.intcon(UINT32_MAX);
+  }
+  Node* pin_count_cmp = _gvn.transform(new CmpUNode(_gvn.transform(pin_count), pin_count_rhs));
+  Node* test_pin_count_over_underflow = _gvn.transform(new BoolNode(pin_count_cmp, BoolTest::eq));
+  IfNode* iff_pin_count_over_underflow = create_and_map_if(control(), test_pin_count_over_underflow, PROB_MIN, COUNT_UNKNOWN);
+
+  // False branch, no pin count over/underflow. Increment or decrement pin count and store back.
+  Node* valid_pin_count = _gvn.transform(new IfFalseNode(iff_pin_count_over_underflow));
+  set_control(valid_pin_count);
+
+  Node* next_pin_count;
+  if (unpin) {
+    next_pin_count = _gvn.transform(new SubINode(pin_count, _gvn.intcon(1)));
+  } else {
+    next_pin_count = _gvn.transform(new AddINode(pin_count, _gvn.intcon(1)));
+  }
+
+  Node* updated_pin_count_memory = store_to_memory(control(), pin_count_offset, next_pin_count, T_INT, Compile::AliasIdxRaw, MemNode::unordered);
+
+  // True branch, pin count over/underflow.
+  Node* pin_count_over_underflow = _gvn.transform(new IfTrueNode(iff_pin_count_over_underflow));
+  {
+    // Deoptimize and reexecute in the interpreter to throw IllegalStateException for pin count over/underflow.
+    PreserveJVMState pjvms(this);
+    PreserveReexecuteState preexecs(this);
+    jvms()->set_should_reexecute(true);
+    set_control(pin_count_over_underflow);
+    set_all_memory(input_memory_state);
+    uncommon_trap_exact(Deoptimization::Reason_intrinsic,
+                        Deoptimization::Action_reinterpret);
```

Comment:
Why do you use `Action_reinterpret` and not `Action_make_not_entrant` here?

You may need a check to avoid endless re-compilation of a method that always hits the trap:
```
  if (too_many_traps(Deoptimization::Reason_intrinsic)) {
    return false;
  }
```

See other intrinsics.

Reply:
A pin_count overflow/underflow should be a per-thread condition, not global.  If there is nothing in the nmethod to be invalidated, maybe this should be `Action_none`?

Reply:
Right, that would make sense to me because `Deoptimization::Action_reinterpret` might also invalidate the nmethod.

Reply:
The functional requirement I have is that the branch takes an uncommon trap and restarts / re-executes the same method the interpreter, because that version enters the VM where an IllegalStateException is thrown. 

I don't need the compiled method to be invalidated, only that an attempt that over/underflows (thread-local) restarts in the interpeter for the exception to be thrown. Is Action_none better suited for this purpose?

Reply:
The pattern of the uncommon trap construct was taken from the precedent in LibraryCallKit::inline_profileBoolean().

Reply:
Is it an implicit invariant that execution always continues in the interpreter after an uncommon trap? I.e., I don't need to explicitly tell it to "re-execute" there?

Reply:
It is updated to use Action::none to keep the nmethod. The trap code picks up the correct bytecode (invokestatic) from the trap scope. So after unrolling the host method (the inliner), the trap bytecode (i.e., the invokestatic call to Continuation.pin() or unpin()) is re-executed in the interpreter. This is also without setting the explicit re-execute state (which may mean something else).

Reply:
Yes, we should not throw out compiled nmethod if one thread need to got into Interpreter and throw an exception.
Other threads will continue to use this nmethod.

Reply:
> This is also without setting the explicit re-execute state (which may mean something else).

I think reexecute is implicitly set for uncommon traps, and the explicit flag is for deoptimization safepoints.